src/jdk.jfr/share/classes/jdk/jfr/internal/StringPool.java

```diff
@@ -71,27 +73,49 @@ private static long externalSid(long internalSid) {
         return internalSid >> 16;
     }
 
-    /* synchronized because of writing the string to the JVM. */
-    private static synchronized long storeString(String s) {
-        Long lsid = cache.get(s);
-        long internalSid;
-        if (lsid != null) {
-            internalSid = lsid.longValue();
-            if (isCurrentGeneration(internalSid)) {
-                // Someone already updated the cache.
-                return externalSid(internalSid);
+    /* Explicitly pin a virtual thread before acquiring the string pool monitor
+     * because migrating the EventWriter onto another carrier thread is impossible.
+     */
+    private static void pinVirtualThread() {
+        if (Thread.currentThread().isVirtual() && ContinuationSupport.isSupported()) {
+            Continuation.pin();
+        }
+    }
+
+    private static void unpinVirtualThread() {
+        if (Thread.currentThread().isVirtual() && ContinuationSupport.isSupported()) {
+            Continuation.unpin();
+        }
+    }
+
+    private static long storeString(String s) {
+        try {
+            pinVirtualThread();
```

Comment:
The pin should be outside the try so that the finally can only happen if pinning was succesful.

Reply:
Good point, thanks.

Reply:
Yes, it should only unpin if pin completed successfully.

Reply:
Please see updated changeset.